File: PrefixSum/3494. Find the Minimum Amount of Time to Brew Potions/solution.py
class Solution:
    def minTime(self, skill: List[int], mana: List[int]) -> int:
        n, m = len(skill), len(mana)
        presum_skill = list(accumulate(skill, initial=0))

        finish_t = [0] * n
        for j in range(m):
            require_t = 0
            for i in range(n):
                # next_finish_t = finish_t[i] + sum(skill[j] for j in range(i, n)) * mana[j]
                next_finish_t = finish_t[i] + (presum_skill[n]-presum_skill[i]) * mana[j]
                require_t = max(require_t, next_finish_t)
            
            finish_t[-1] = max(finish_t[-1], require_t)
            for i in range(n-2, -1, -1):
                finish_t[i] = finish_t[i+1] - skill[i+1] * mana[j]

        return finish_t[-1]

File: PrefixSum/3494. Find the Minimum Amount of Time to Brew Potions/solution.cpp
class Solution {
public:
    long long minTime(vector<int>& skill, vector<int>& mana) {
        if (skill.empty() || mana.empty()) return 0;

        int n = skill.size(), m = mana.size();
        
        // Compute prefix sum of skills
        vector<long long> prefix_sum(n);
        partial_sum(skill.begin(), skill.end(), prefix_sum.begin());

        // Track completion times for each worker
        vector<long long> finish_times(n, 0);// Process each task
        for (int j = 0; j < m; j++) {
            long long required_time = 0;
            // Calculate max time for task j
            for (int i = 0; i < n; i++) {
                long long remaining_skill = prefix_sum[n-1] - (i > 0 ? prefix_sum[i-1] : 0);
                long long next_finish_time = finish_times[i] + remaining_skill * mana[j];
                required_time = max(required_time, next_finish_time);
            }
            
            // Update finish time for the last worker
            finish_times[n-1] = max(finish_times[n-1], required_time);
            
            // Backpropagate finish times
            for (int i = n-2; i >= 0; i--) {
                finish_times[i] = finish_times[i+1] - skill[i+1] * mana[j];
            }
        }
        return finish_times[n-1];
    }
};